### ISSUE TITLE
Bug: [restrict-template-expressions] Wrong "Invalid type "any" of template literal expression." message

### ISSUE THREAD

**Issue by viceice** [typescript-eslint/typescript-eslint#5199]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=C4TwDgpgBAShB2B7AbgQ2BAwo+AzAlgOZQC8UA3gL4CwAUKJFALJEBO6+OpFNt+8GVrlQBjaC0LtgneNngBnYKwCuI4IlYU6UKPAgB3ABTadUDUX6oANnIKEAXLAQp0WHHYA0JnQFs2rgBNbIkc4JDQMYMITAEpHCSkZAG46XjoAiBErVFZoLNR5eWZ-aRx5AGUIVmR8MS1aHUUOESh8eQTAwz9JQKjQ5wi3PCIPM1YLeGs+p3DXKLioACNERCsIVHgUhqgm6RaVeENzQksbdxCZl0jzwgWwq6G7LZ0wVkQMNQgAneBmqEIIMAOqUFMZtjpjqd+rNrsNCF5wVBulIvtCHlFYv1UAEcFYQABBVjsEAAHmBMgAfFs0rQMvlclB8oUoBAAB6QNT1Rq-PYs1kYeABLroEQACyq8niIvFrHkAHlFgArTLABbIRD4ALUug6+jgcTSiUK5WcsjkEwAbUmPggjkU43ghAAuqFUPomIbWAAxTapLZ0BjQGBuj3AMVVH0k8o86BsgUBIqh8OsaOubhJmWpjAU7jm7aGYCitqOLMQUaiYDKayODYgUYAOkbbI5GACjkJxJJtYpCwAouyVRwcHB5MorMBqf69YwM1VS7mTBAAI5VqySqBg0xQVCOZTwADWSH08ARW8Wu4PR5P3kZykUiB8ABUIIoJQB+dtE1CkveHxDHilT1Me1amATBxREfcPyWFY1g2ICYlIHNllWdZfVoXhAygfsWyHeARzHYBcxpZsVXrOMECFPMdHUYEIE3UxMDvdQfHJHB4hKGQ5HtVR1FYICIXGE5JjOOE0UGKIBL5Fsvmme4JJuKS2jo74yCUZQIFiLlTCZIomPvVjOLKSpqlqWN+UoxMjIUEyajqaitygV53hVL4fj+FAqnGDJ-kBNjQRvLdIRE8S5kUwLTGRQJQthOwIrudYcXgPEO2-MlrJzBzHKgXJK1YeAoCtAwoH0lj-KOITTlGKLWxiJ1nkc3gtxpLcRDKIiatkm5uFKh9-IqKo7IgesDgqiYphuGIpy3fBcA3CL+tssz62Ukovi6NagkUsZxtEuxEIAQhIEgItWnpWxvRCstMXLlHy7TsptQpUABRxDAWEDHSQqAAAMzpRb55FFRAx2+RZoAAEnIf7AkoH6pK3MACnXYQ1zLCLKAanQWtMWaNwOws2nIldrHkDbzq6uFRlIz4trhGIroi277uurcnvkF7bQ3D6lH4YgSApCKtx+-yoGEfA1gCAAdeAZbgMR8GQNy2rEmWoYAKXKOUADl60+k5cBABjsrZzaMURbKYkoGWZZw1zvhVux7DV8hNZ1vXecdWajZp1t5jhhHTCRwpHFR+R0YtqBMZvHGdGZgrWaRF8Ode7m7U9-mcwAclF4Pw8B1QxEKXAxzxLPA7zxx1Ij5qGsoBFKCmuggA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aFfKsu-RAW2LwAhn3oAPYh2TJKAeyaoMAbXDYIAJgA0a9VnXZIw+PDkB3AHKx+AI0TR0YLrEQ6Dh46bMAhOXKTCTI7OiLrYAL5hALpqkeFAA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
type RenovateConfig = {}
type Migration = {}
interface MigrationConstructor {
  new(
    originalConfig: RenovateConfig,
    migratedConfig: RenovateConfig
  ): Migration;
}

declare class MigrationsService {
  static isMigrated(migratedConfig: RenovateConfig, originalConfig: RenovateConfig): boolean;
  static run(originalConfig: RenovateConfig): RenovateConfig;
  protected static getMigrations(
    original: RenovateConfig,
    migrated: RenovateConfig
  ): ReadonlyArray<Migration>;
}

declare class expect {
  static extend(matchers: MatchersObject): void;
}


type MatchersObject = {
  [name: string]: RawMatcherFn;
};

type RawMatcherFn<State extends MatcherState = MatcherState> = {
  (this: State, actual: any, ...expected: Array<any>): ExpectationResult;
};

type MatcherState = {
  equals: (
    a: unknown,
    b: unknown,
    customTesters?: Array<unknown>,
    strictCheck?: boolean,
  ) => boolean;
}
type ExpectationResult = {}

expect.extend({
  toMigrate(
    CustomMigration: MigrationConstructor,
    originalConfig: RenovateConfig,
    expectedConfig: RenovateConfig,
    isMigrated = true
  ) {
    class CustomMigrationsService extends MigrationsService {
      protected static override getMigrations(
        original: RenovateConfig,
        migrated: RenovateConfig
      ): ReadonlyArray<Migration> {
        return [new CustomMigration(original, migrated)];
      }
    }

    const migratedConfig = CustomMigrationsService.run(originalConfig);

    if (
      MigrationsService.isMigrated(migratedConfig, originalConfig) !==
      isMigrated
    ) {
      return {
        message: (): string => `isMigrated should be ${isMigrated}`,
        pass: false,
      };
    }

    if (!this.equals(migratedConfig, expectedConfig)) {
      return {
        message: (): string =>
          `Migration failed\n\nReceived config:\n${JSON.stringify(
            migratedConfig
          )}\n\nExpected config:\n${JSON.stringify(expectedConfig)}`,
        pass: false,
      };
    }

    return {
      message: (): string => 'Migration passed successfully',
      pass: true,
    };
  },
});
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/restrict-template-expressions": [
      2,
      {
        "allowNumber": true,
        "allowBoolean": true
      }
    ]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

There should be no error.

### Actual Result

it shows ' Invalid type "any" of template literal expression.'  error

### Versions

| package | version |
| -- | -- |
| `@typescript-eslint/eslint-plugin` | `5.28.0` |
| `@typescript-eslint/parser` | `5.28.0` |
| `TypeScript` | `4.7.3` |
| `ESLint` | `8.17.0` |
| `node` | `web` |

**Comment by viceice** [comment]
happens here: https://github.com/renovatebot/renovate/blob/main/test/to-migrate.ts

**Comment by viceice** [comment]
Probably again a false positive because of wrong type inference

**Comment by Josh-Cena** [comment]
This should be a `no-inferrable-types` bug. The parameter type is not inferable in a callback because there's the contextual type as inference source, which overrides inference from the default parameter.

**Comment by bradzacher** [comment]
![image](https://user-images.githubusercontent.com/7462525/174123641-2ff041d9-f1af-4530-9000-53a2c885eb9a.png)

This is working as expected - TS thinks the type of `isMigrated` is `any`.

> This should be a no-inferrable-types bug

It's impossible to tell this without types, and even with types it's **hard**.
#4273, #3752, #3145, #2184
Which is why we have the [`ignoreParameters`](https://typescript-eslint.io/rules/no-inferrable-types/#ignoreparameters) option for the rule, in case your codebase has lots of unsafely typed code that requires an annotation.

**Comment by viceice** [comment]
so the only options are?
1. extract `toMigrate` as ordinary function
2. use `ignoreParameters` and explicit annotate `isMigrated`

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=C4TwDgpgBAShB2B7AbgQ2BAwo+AzAlgOZQC8UA3gL4CwAUKJFALJEBO6+OpFNt+8GVrlQBjaC0LtgneNngBnYKwCuI4IlYU6UKPAgB3ABTadUDUX6oANnIKEAXLAQp0WHHYA0JnQFs2rgBNbIkc4JDQMYMITAEpHCSkZAG46XjoAiBErVFZoLNR5eWZ-aRx5AGUIVmR8MS1aHUUOESh8eQTAwz9JQKjQ5wi3PCIPM1YLeGs+p3DXKLioACNERCsIVHgUhqgm6RaVeENzQksbdxCZl0jzwgWwq6G7LZ0wVkQMNQgAneBmqEIIMAOqUFMZtjpjqd+rNrsNCF5wVBulIvtCHlFYv1UAEcFYQABBVjsEAAHmBMgAfFs0rQMvlclB8oUoBAAB6QNT1Rq-PYs1kYeABLroEQACyq8niIvFrHkAHlFgArTLABbIRD4ALUug6+jgcTSiUK5WcsjkEwAbUmPggjkU43ghAAuqFUPomIbWAAxTapLZ0BjQGBuj3AMVVH0k8o86BsgUBIqh8OsaOubhJmWpjAU7jm7aGYCitqOLMQUaiYDKayODYgUYAOkbbI5GACjkJxJJtYpCwAouyVRwcHB5MorMBqf69YwM1VS7mTBAAI5VqySqBg0xQVCOZTwADWSH08ARW8Wu4PR5P3kZykUiB8ABUIIoJQB+dtE1CkveHxDHilT1Me1amATBxREfcPyWFY1g2ICYlIHNllWdZfVoXhAygfsWyHeARzHYBcxpZsVXrOMECFPMdHUYEIE3UxMDvdQfHJHB4hKGQ5HtVR1FYICIXGE5JjOOE0UGKIBL5Fsvmme4JJuKS2jotsYNQjZuCUZQIFiLlTCZIomPvVjOLKSpqlqWN+UoxNTIUcyajqaitygV53hVL4fj+FAqnGDJ-kBNjQRvLdIRE8S5kUkLTGRQIIthOxorudYcXgPEO2-Mk7JzZyXKgXJK1YeAoCtAwoCMligqOITTlGWLWxiJ1nhc3gtxpLcRDKIj6tkm5uAqh8goqKpHIgesDmqiYphuGIpy3fBcA3aKhocyz62Ukovi6TagkUsYptEuxEIAQhIEhoo2npWxvRDctMArlCKvS8ptQpUABRxDAWEDHSQqAAANLpRb55FFRAx2+RZoAAEnIIHAkof6pK3MACnXYQ1zLaLKGanR2tMBaN2Ows2nIldrHkbart6uFRlIz5drhGJbuih6nrurdXvkd7bQ3b6lH4YgSApaKt3+oKoGEfA1gCAAdeB5bgMR8GQTzOrE+XYYAKXKOUADl6x+k5cBABi8s5naMURPKYkoeX5ZwjzvnVux7E18gdf1w2BcdBbTfp1t5kR5HTFRwpHAx+QsetqAcZvfGdDZ4qOaRF9uY+vm7R9oWcwAcglsOo5B1QxEKXAxzxXOQ8LxwtOjtrmsoBFKFmuggA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aFfKsu-RAW2LwAhn3oAPYh2TJKAeyaoMAbXDYIAJgA0a9VnXZIw+PDkB3AHKx+AI0TR0YLrEQ6Dh46bMAhOXKTCTI7OiLrYAL5hALpuhgQk5FQ09MiMLKxMcrTMAGb20MI2SLREpEqQ+XIOapHhQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

**Comment by bradzacher** [comment]
Alternatively you can just use a disable comment and detail why you're working around the rule.
